# datetime
---


In [21]:
import datetime

In [3]:
print(dir(datetime))

['MAXYEAR', 'MINYEAR', 'UTC', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'date', 'datetime', 'datetime_CAPI', 'time', 'timedelta', 'timezone', 'tzinfo']


## Time

In [4]:
# datetime.time(hour=0, minute=0, second=0, microsecond=0, tzinfo=None, *, fold=0)

t = datetime.time(1, 2, 3)
print(t)

01:02:03


In [5]:
print(f'hour: {t.hour}')
print(f'minute: {t.minute}')
print(f'second: {t.second}')
print(f'microsecond: {t.microsecond}')
print(f'tzinfo: {t.tzinfo}')

hour: 1
minute: 2
second: 3
microsecond: 0
tzinfo: None


In [6]:
# variable 't' is a time object, not a datetime object
print('Earliest:', datetime.time.min)
print('Latest:', datetime.time.max)
print('Resolution:', datetime.time.resolution)

Earliest: 00:00:00
Latest: 23:59:59.999999
Resolution: 0:00:00.000001


In [7]:
midnigth = datetime.time(0, 39, 0, 5)
midnigth2 = datetime.time(microsecond=5, minute=39)
print(midnigth)
print(midnigth2)

00:39:00.000005
00:39:00.000005


## Date

In [8]:
# datetime.date(year, month, day)

today = datetime.date.today()
print(f'today: \n\t {today}')
# ctime is a string representation of the date
print(f'ctime: \n\t {today.ctime()}')
# tuple is a tuple representation of the date
print(f'tuple: \n\t {today.timetuple()}')
# ordinal is a number of days since 1/1/1
print(f'ordinal: \n\t {today.toordinal()}')
print(f'year: \n\t {today.year}')
print(f'month: \n\t {today.month}')
print(f'day: \n\t {today.day}')

today: 
	 2024-12-23
ctime: 
	 Mon Dec 23 00:00:00 2024
tuple: 
	 time.struct_time(tm_year=2024, tm_mon=12, tm_mday=23, tm_hour=0, tm_min=0, tm_sec=0, tm_wday=0, tm_yday=358, tm_isdst=-1)
ordinal: 
	 739243
year: 
	 2024
month: 
	 12
day: 
	 23


In [9]:
import time

o = 733114
print(f'o = {o} and fromordinal(o) = {datetime.date.fromordinal(o)}')

t = time.time()
print(f't = {t} and fromtimestamp(t) = {datetime.date.fromtimestamp(t)}')

o = 733114 and fromordinal(o) = 2008-03-13
t = 1734952768.2622771 and fromtimestamp(t) = 2024-12-23


In [10]:
print('Earliest:', datetime.date.min)
print('Latest:', datetime.date.max) 
print('Resolution:', datetime.date.resolution) # whole day

Earliest: 0001-01-01
Latest: 9999-12-31
Resolution: 1 day, 0:00:00


In [11]:
d1 = datetime.date(2022, 7, 5)
print(f'd1: {d1}')

d2 = d1.replace(year=2023)
print(f'd2: {d2}')

d1: 2022-07-05
d2: 2023-07-05


In [12]:
help(d1.replace)

Help on built-in function replace:

replace(...) method of datetime.date instance
    Return date with new specified fields.



In [13]:
created_date = datetime.date(year=2023, month=1, day=5)
print(created_date)

2023-01-05


In [24]:
# create a date where year is 2023 and the day is 5

d = datetime.date(day=5, month=1, year=2023)
print(d)

2023-01-05


## timedeltas

In [14]:
# datetime.timedelta(days=0, seconds=0, microseconds=0, milliseconds=0, minutes=0, hours=0, weeks=0)

print(f'microseconds: {datetime.timedelta(microseconds=1)}')
print(f'milliseconds: {datetime.timedelta(milliseconds=1)}')
print(f'seconds: {datetime.timedelta(seconds=1)}')
print(f'minutes: {datetime.timedelta(minutes=1)}')
print(f'hours: {datetime.timedelta(hours=1)}')
print(f'days: {datetime.timedelta(days=1)}')
print(f'weeks: {datetime.timedelta(weeks=1)}')

microseconds: 0:00:00.000001
milliseconds: 0:00:00.001000
seconds: 0:00:01
minutes: 0:01:00
hours: 1:00:00
days: 1 day, 0:00:00
weeks: 7 days, 0:00:00


### Date Arithmetic
Date math uses the standard arithmetic operators.


The following example with date objects illustrates using timedelta objects to compute new dates, and subtracting date instances to produce timedeltas (including a negative delta value).

In [15]:
today = datetime.date.today()
print(f'today: {today}')

today: 2024-12-23


In [16]:
one_timedelta_day = datetime.timedelta(days=1)
print(f'one_timedelta_day: {one_timedelta_day}')

one_timedelta_day: 1 day, 0:00:00


In [17]:
yesterday = today - one_timedelta_day
print(f'yesterday: {yesterday}')

yesterday: 2024-12-22


In [18]:
tomorrow = today + one_timedelta_day
print(f'tomorrow: {tomorrow}')

tomorrow: 2024-12-24


In [19]:
print(f'tomorrow - yesterday: {tomorrow - yesterday}')
print(f'yesterday - tomorrow: {yesterday - tomorrow}')

tomorrow - yesterday: 2 days, 0:00:00
yesterday - tomorrow: -2 days, 0:00:00


In [25]:
# Create an arbitary date and add 5 weeks to it

date = datetime.date.today()
print(date)
new_date = date + datetime.timedelta(weeks=5)
print(new_date)

2024-12-23
2025-01-27


## Comparin Values

both date and time values can be compared using the standart operators to determine whis is earlier or later

In [26]:
print('Times:')
t1 = datetime.time(12,55,0)
print ('\tt1:', t1)
t2 = datetime.time(13,5,0)
print('\tt2:', t2)
print('\tt1 < t2:', t1 < t2)

Times:
	t1: 12:55:00
	t2: 13:05:00
	t1 < t2: True


In [27]:
print ('Dates:')
d1 = datetime.date.today()
print('\td1:', d1)
d2 = datetime.date.today() + datetime.timedelta(days=1)
print('\td2:', d2)
print('\td1 > d2:', d1 > d2)

Dates:
	d1: 2024-12-23
	d2: 2024-12-24
	d1 > d2: False


## Combining Dates and Times: datetime Class

- we can use datetime class to hold values consisting of both date and time components
- Represents all the attributes from date and time objects

```python
datetime.datetime(year, month, day,
                    hour= 0, minute=0, second=0,
                    microsecond=0, tzinfo=None, *, fold=0)
```
more information in the [document](https://docs.python.org/3/library/datetime.html)

In [29]:
print(f'Now: {datetime.datetime.now()}')
print(f'Today: {datetime.datetime.today()}')
print(f'UTC Now: {datetime.datetime.now(datetime.UTC)}')


Now: 2024-12-23 13:22:05.961900
Today: 2024-12-23 13:22:05.962991
UTC Now: 2024-12-23 13:22:05.963049+00:00


In [30]:
d = datetime.datetime.now()
for attr in ['year', 'month', 'day', 'hour', 'minute', 'second', 'microsecond']:
    print(f'{attr}: {getattr(d, attr)}')

year: 2024
month: 12
day: 23
hour: 13
minute: 23
second: 3
microsecond: 163971


In [32]:
NOW = datetime.datetime.now()

print(f'Current date & time:         {NOW}')
print(f'Date and Time in ISO format: {NOW.isoformat()}')
print(f'Current year:                {NOW.year}')
print(f'Current month:               {NOW.month}')
print(f'Current day:                 {NOW.day}')
print(f'dd/mm/yyyy format:           {NOW.strftime("%d/%m/%Y")}')
print(f'Current hour:                {NOW.hour}')
print(f'Current minute:              {NOW.minute}')
print(f'Current second:              {NOW.second}')
print(f'hh:mm:ss format:             {NOW.strftime("%H:%M:%S")}')

Current date & time:         2024-12-23 13:26:22.266174
Date and Time in ISO format: 2024-12-23T13:26:22.266174
Current year:                2024
Current month:               12
Current day:                 23
dd/mm/yyyy format:           23/12/2024
Current hour:                13
Current minute:              26
Current second:              22
hh:mm:ss format:             13:26:22


- just as with date, datetime provides convenient class methods for creating new instance. it also inclides `fromordinal()` and `fromtimestamp()`.
- In addition, `combine()` can be useful if you already have a date instance and want to create a datetime. 

In [33]:
t = datetime.time(1,2,3)
print('t:', t)

d = datetime.date.today()
print('d:', d)

dt = datetime.datetime.combine(d,t)
print('dt:', dt)

t: 01:02:03
d: 2024-12-23
dt: 2024-12-23 01:02:03


## Formatting and Parsing

- The default string representation of a datetime object uses the format YYY-MM-DDTHH:MM:SS.mmmmmm
- Alternate formats can be generated using `strftime()` function.
    - Converts datetime objects into string
    - Uses different control code to give an output.
    - Each control code resembles different parameters like year, month, weekday and date.
- If your input data includes timestamp values parsable with `time.strptime()`, then `datetime.strptime()` is a convenient way to convert them to datetime instance. 

**Useful `strptime` and `strftime` Patterns**\n
    
|Directive | Meaning |
| --- | --- |
| `%a` | Weekday as locale's abbreviated name |
| `%A` | Weekday as locale's full name |
| `%w` | Weekday as decimal number, where 0 is Sunday and 6 is Saturday |
| `%d` | Day of the month as a zero-padded decimal number [01, 02, ..., 30, 31] |
| `%b` | Month as locale's abbreviated name |
| `%B` | Weekday as locale's full name |
| `%m` | Month as zero-padded decimal number [01, 02, ..., 11, 12] |
| `%y` | Year without century as a zero-padded decimal number name [0001, ..., 2019, 2020, ..., 9999] |
| `%Y` | Year with century as a decimal number  |
| `%H` | Hour (24-hour clock) as a zero-padded decimal number [00, 01, ..., 22, 23] |
| `%I` | Hour (12-hour clock) as a zero-padded decimal number |
| `%p` | Locale equivalent of either AM or PM |
| `%M` | Minute as a zero-padded decimal number [00, 01, ..., 58, 59]|
| `%S` | Second as a zero-padded decimal number [00, 01, ..., 58, 59]|
| `%f` | Microsecond as a zero-padded decimal number |
| `%j` | Day of the year as a zero-padded decimal number |
| `%W` | Week number of the year (Monday as the first day of the week) as a decimal number |
| `%U` | Week number of the year (Sunday as the first day of the week) as a decimal number |
| `%c` | Locale’s appropriate date and time representation |
| `%Z` | Time zone name |
| `%z` | UTC offset in the form HH[SS[.fffff]] 

### Formatting

Weekday, Month, Day, Hour, Minute, Seconds, Year

In [34]:
format = "%a %d %b %H:%M:%S %Y"

today = datetime.datetime.today()
print(f'ISO: {today}')

s = today.strftime(format)
print(f'strftime: {s}')

ISO: 2024-12-23 16:24:47.362920
strftime: Mon 23 Dec 16:24:47 2024


In [35]:
# Obtain the time in HH:MM:SS format
print(today.strftime('%X'))

16:24:47


In [36]:
# Obtain the hour with 12 hours time format
print(today.strftime('%I'))

04


In [37]:
# Obtain AM and PM 
print(today.strftime('%p'))

PM


In [38]:
# %c - local date and time, %x local date, %X local time
print(f' Date and Time: {today.strftime("%c")}')
print(f' Date:          {today.strftime("%x")}')
print(f' Time:          {today.strftime("%X")}')

 Date and Time: Mon Dec 23 16:24:47 2024
 Date:          12/23/24
 Time:          16:24:47


In [39]:
# %I/%H -12/24 Hour, %M - minute, %S - second, %p - local's AM/PM
print(f' Time:       {today.strftime("%I:%M:%S %p")}')
print(f'Hour:Minute: {today.strftime("%H:%M")}')

 Time:       04:24:47 PM
Hour:Minute: 16:24


### Parsing

If the format of a string is known, it can be easily parsed to a `datetime` object using `strptime()`

In [40]:
datetime_str = 'Sep 19 2020 2:42PM'
datetime_fmt = '%b %d %Y %I:%M%p'
datetime_obj = datetime.datetime.strptime(datetime_str, datetime_fmt)

print('Date:',      datetime_obj.date())
print('Time:',      datetime_obj.time())
print('Date-time:', datetime_obj)

Date: 2020-09-19
Time: 14:42:00
Date-time: 2020-09-19 14:42:00


In [41]:
print('String date:', datetime_obj.strftime(datetime_fmt))

String date: Sep 19 2020 02:42PM


#### Breakout 4

- Write a function named `number_days_between` that:
    - Takes two argument that are 8-digit integer of the form YYYYMMDD (actual date) 
    - Returns the number of days between the two dates

- For instance:
```python
    number_days_between(20200617, 20200619) = 2
    number_days_between(20200619, 20100219) = 3773
```

In [49]:
# number of days between two dates function, it takes two integer in the format of YYYYMMDD and returns the number of days between them
def days_between_dates(date1, date2):
    date_format = '%Y%m%d'
    d1 = datetime.datetime.strptime(str(date1), date_format)
    d2 = datetime.datetime.strptime(str(date2), date_format)
    
    if d1 < d2:
        return (d2 - d1).days
    else:
        return (d1 - d2).days

number_of_days = days_between_dates(20220101, 20220131)
print(f'Number of days between 20220101 and 20220131: {number_of_days}')

Number of days between 20220101 and 20220131: 30
